activerecord/lib/active_record/autosave_association.rb

```diff
@@ -329,6 +329,7 @@ def validate_collection_association(reflection)
       # enabled records if they're marked_for_destruction? or destroyed.
       def association_valid?(reflection, record, index = nil)
         return true if record.destroyed? || (reflection.options[:autosave] && record.marked_for_destruction?)
+        return true if !custom_validation_context? && record.persisted? && !record.has_changes_to_save? && !record.valid?
```

Comment:
The autosave code is very complex with a lot of conditionals.
I'm wondering if this should be handled instead in https://github.com/rails/rails/blob/34b905d082483e343c793df7b03597a45c86b249/activerecord/lib/active_record/autosave_association.rb#L279-L287

Reply:
I think it has to be done in `association_valid?` because we need to skip validation but keep it for saving.

I noticed that `!record.valid?` is not necessary.

Maybe that helps?